### ISSUE TITLE
Training problem about QWEN-srpo

### ISSUE THREAD

**Issue by tianqyun111** [Tencent-Hunyuan/SRPO#34]
Thank you for your hard work. I hope to support qwen's training. After modifying the code myself, I found that the memory usage increased sharply during loss.backward(), and it also caused OOM on A100 and 80G. What is the reason for this?
this is the inversion part of my code

f inversion ==0:
                with torch.amp.autocast('cuda',dtype=torch.bfloat16):
                    sigma = sigmas_l[t_base]
                    timestep_value = int(sigma * 1000)
                    timesteps = torch.full([text_embeddings_gpu.shape[0]], timestep_value, device=latent_start.device, dtype=torch.long)
                    # with transformer.cache_context("cond"):
                    noise_pred = transformer(
                        hidden_states=latents,
                        timestep=timesteps / 1000,
                        # guidance=guidance,
                        encoder_hidden_states_mask=attention_mask_gpu,
                        encoder_hidden_states=text_embeddings_gpu,
                        img_shapes=img_shapes,
                        txt_seq_lens=txt_seq_lens,
                        guidance = guidance,
                        attention_kwargs={},
                        return_dict=False,
                    )[0]
                    # if do_true_cfg:
                    # with torch.amp.autocast('cuda',dtype=torch.bfloat16):
                    # with transformer.cache_context("uncond"):
                    neg_noise_pred = transformer(
                        hidden_states=latents,
                        timestep=timesteps / 1000,
                        guidance=guidance,
                        encoder_hidden_states_mask=negative_prompt_embeds_mask,
                        encoder_hidden_states=negative_prompt_embeds,
                        img_shapes=img_shapes,
                        txt_seq_lens=negative_txt_seq_lens,
                        attention_kwargs={},
                        return_dict=False,
                    )[0]
                    comb_pred = neg_noise_pred + true_cfg_scale * (noise_pred - neg_noise_pred)

                    cond_norm = torch.norm(noise_pred, dim=-1, keepdim=True)
                    noise_norm = torch.norm(comb_pred, dim=-1, keepdim=True)
                    pred = comb_pred * (cond_norm / noise_norm)
                dsigma = sigma-sigmas_l[t_start]
                latents = latents.to(torch.bfloat16) - dsigma * pred

            ############Direct-Ailgn step2 denoise one step###############
            else:
                with torch.amp.autocast('cuda',dtype=torch.bfloat16):
                    sigma = sigmas_l[t_start]
                    timestep_value = int(sigma * 1000)
                    timesteps = torch.full([text_embeddings_gpu.shape[0]], timestep_value, device=latent_start.device, dtype=torch.long)
                    # with transformer.cache_context("cond"):
                    noise_pred = transformer(
                        hidden_states=latents,
                        timestep=timesteps / 1000,
                        # guidance=guidance,
                        encoder_hidden_states_mask=attention_mask_gpu,
                        encoder_hidden_states=text_embeddings_gpu,
                        img_shapes=img_shapes,
                        txt_seq_lens=txt_seq_lens,
                        guidance = guidance,
                        attention_kwargs={},
                        return_dict=False,
                    )[0]
                    # if do_true_cfg:
                    # with torch.amp.autocast('cuda',dtype=torch.bfloat16):
                    # with transformer.cache_context("uncond"):
                    neg_noise_pred = transformer(
                        hidden_states=latents,
                        timestep=timesteps / 1000,
                        guidance=guidance,
                        encoder_hidden_states_mask=negative_prompt_embeds_mask,
                        encoder_hidden_states=negative_prompt_embeds,
                        img_shapes=img_shapes,
                        txt_seq_lens=negative_txt_seq_lens,
                        attention_kwargs={},
                        return_dict=False,
                    )[0]
                    comb_pred = neg_noise_pred + true_cfg_scale * (noise_pred - neg_noise_pred)

                    cond_norm = torch.norm(noise_pred, dim=-1, keepdim=True)
                    noise_norm = torch.norm(comb_pred, dim=-1, keepdim=True)
                    pred = comb_pred * (cond_norm / noise_norm)
                dsigma = sigmas_l[t_base] - sigma
                latents = latents.to(torch.bfloat16) + dsigma * pred
            
            # 恢复图像并计算奖励
            latents = (latents - gt_vector) / (1 - sigmast)  # 撤销噪声影响，获得此step的最终latent
            vae.enable_tiling()
            latents = latents.to(device_cpu)
            # with torch.autocast("cuda", dtype=torch.bfloat16):
            latents = unpack_latents(latents, height, width, vae_scale_factor)
            latents = latents.to(vae.dtype)
            latents_mean = (
                torch.tensor(vae.config.latents_mean)
                .view(1, vae.config.z_dim, 1, 1, 1)
                .to(latents.device, latents.dtype)
            )
            latents_std = 1.0 / torch.tensor(vae.config.latents_std).view(1, vae.config.z_dim, 1, 1, 1).to(
                latents.device, latents.dtype
            )
            latents = latents / latents_std + latents_mean
            
            img = vae.decode(latents, return_dict=False)[0][:, :, 0]
            # with torch.no_grad():
            with torch.amp.autocast('cuda',dtype=torch.bfloat16):
                if inversion:
                    reward_score = reward_model.SRP_cfg([pos_caption], batch_caption, img, discount[mid_timestep])
                else:
                    reward_score = reward_model.SRP_cfg(batch_caption, [pos_caption], img, discount_inversion[mid_timestep])
            del img,latents,vae,transformer
            gpu_mem("after reward_score")
            torch.cuda.empty_cache()
            # 与阈值比较计算损失并反向传播
            # loss = F.relu(-reward_score + 0.7) / gradient_accumulation_steps
            loss = F.relu(-reward_score + 0.7) / gradient_accumulation_steps
            loss = loss.mean()
            gpu_mem("before loss.backward()")
            loss.backward()

**Comment by tianqyun111** [comment]
> The likely reason is related to the CFG. Flux uses a single forward pass with gradients (true_cfg not needed), but your code runs two forward passes per step, doubling memory usage during loss.backward(). In our practice, setting true_cfg=1 (no neg_noise_pred = transformer) greatly reduces memory consumption while still achieving reasonable results or decrease the h, w for training. Also, this code hasn't been fully optimized with acceleration frameworks yet—you may consider integrating DeepSpeed or similar to save memory.

i tried to bypass the neg_noise_pred,but it still showed oom when loss.backward(),so i think it is necessary to integrating deepspeed.
BTW,when can you release qwen-srpo version?I eagerly await your progress！！！

**Comment by nitinh12** [comment]
> > The likely reason is related to the CFG. Flux uses a single forward pass with gradients (true_cfg not needed), but your code runs two forward passes per step, doubling memory usage during loss.backward(). In our practice, setting true_cfg=1 (no neg_noise_pred = transformer) greatly reduces memory consumption while still achieving reasonable results or decrease the h, w for training. Also, this code hasn't been fully optimized with acceleration frameworks yet—you may consider integrating DeepSpeed or similar to save memory.
> 
> i tried to bypass the neg_noise_pred,but it still showed oom when loss.backward(),so i think it is necessary to integrating deepspeed. BTW,when can you release qwen-srpo version?I eagerly await your progress！！！

I am also waiting for the Qwen SRPO model. Have you tried to train on RTX PRO 6000? If you can share your code, I can train it.
